handler/src/main/java/io/netty/handler/ssl/JdkApplicationProtocolNegotiator.java

```diff
@@ -17,120 +17,115 @@
 
 import io.netty.buffer.ByteBufAllocator;
 import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLHandshakeException;
 import java.util.List;
-import java.util.Set;
+
+import io.netty.handler.ssl.ApplicationProtocolConfig.Protocol;
+import io.netty.util.internal.PlatformDependent;
 
 /**
  * JDK extension methods to support {@link ApplicationProtocolNegotiator}
  */
-public interface JdkApplicationProtocolNegotiator extends ApplicationProtocolNegotiator {
-    /**
-     * Abstract factory pattern for wrapping an {@link SSLEngine} object. This is useful for NPN/APLN JDK support.
-     */
-    interface SslEngineWrapperFactory {
-        /**
-         * Abstract factory pattern for wrapping an {@link SSLEngine} object. This is useful for NPN/APLN support.
-         *
-         * @param engine The engine to wrap.
-         * @param alloc the buffer allocator.
-         * @param applicationNegotiator The application level protocol negotiator
-         * @param isServer <ul>
-         * <li>{@code true} if the engine is for server side of connections</li>
-         * <li>{@code false} if the engine is for client side of connections</li>
-         * </ul>
-         * @return The resulting wrapped engine. This may just be {@code engine}.
-         */
-        SSLEngine wrapSslEngine(SSLEngine engine, ByteBufAllocator alloc,
-                JdkApplicationProtocolNegotiator applicationNegotiator, boolean isServer);
-    }
+class JdkApplicationProtocolNegotiator implements ApplicationProtocolNegotiator {
+    private static final boolean ALPN_AVAILABLE = Conscrypt.isAvailable()
+            || JettyAlpnSslEngine.isAvailable()
+            || PlatformDependent.javaVersion() >= 9;
+    private static final boolean NPN_AVAILABLE = JettyNpnSslEngine.isAvailable();
+    static final JdkApplicationProtocolNegotiator DEFAULT =
+            new JdkApplicationProtocolNegotiator(null, false, Protocol.NONE);
+
+    private final List<String> supportedProtocols;
+    private final boolean failIfNoCommonProtocols;
+    private final Protocol protocol;
+
+    JdkApplicationProtocolNegotiator(
+        List<String> supportedProtocols,
+        boolean failIfNoCommonProtocols,
+        Protocol protocol
+    ) {
+        this.supportedProtocols = supportedProtocols;
```

Comment:
wraps this with `Collections.unmodifiableList(...)` as we return the List via a getter.